### PR TITLE
Feature/passing direct url and local path to agent runner epmdj 10700

### PR DESCRIPTION
**Pull request by bodyangug**

### CODE REVIEW

File: common/src/main/kotlin/AgentLoader.kt [hunk 1]
```diff
@@ -23,15 +23,40 @@ object AgentLoader {
...
+
+        // Case 3. None of the props are setting-up
+        if (directUrlToZip == null && directLocalPathToZip == null) {
+            val artPath = "${config.repositoryName}/releases/download/v${config.version}/$artName"
+            val urlToDownload = "${GITHUB_URL.replace("api.", "")}/$DRILL_OWNER/$artPath"
+            file = dir.resolve(artName)
+            file.ensureFile(URL(urlToDownload))
+        }
         unzip(file, dir)
     }
```
Comment (RomanDavlyatshin) [inline common/src/main/kotlin/AgentLoader.kt:51]:
```
    fun downloadAgent(
        config: Configuration,
        dir: File
    ) = run {
        var file: File
        val artName = "${config.artifactId}-$presetName-${config.version}.zip"

        if (config.directLocalPathToZip != null) {
            println("Download agent: from local path ${config.directLocalPathToZip}")
            file = dir.resolve(config.directLocalPathToZip)
        } else if (config.directUrlToZip != null) {
            println("Download agent: from direct url ${config.directUrlToZip}")
            file = dir.resolve(artName)
            file.ensureFile(URL(config.directUrlToZip))
        } else {
            val artPath = "${config.repositoryName}/releases/download/v${config.version}/$artName"
            val urlToDownload = "${GITHUB_URL.replace("api.", "")}/$DRILL_OWNER/$artPath"
            println("Download agent: from default repository $urlToDownload")
            file = dir.resolve(artName)
            file.ensureFile(URL(urlToDownload))
        }

        unzip(file, dir)
    }
```
TODO - add error handling

File: common/src/main/kotlin/AgentLoader.kt [hunk 2]
```diff
@@ -23,15 +23,40 @@ object AgentLoader {
...
 
+    private fun File.ensureFile(downloadFrom: URL) {
```
Comment (RomanDavlyatshin) [inline common/src/main/kotlin/AgentLoader.kt:53]: Misleading function name - it also downloads file

File: common/src/main/kotlin/AgentLoader.kt (hunk 1 above)
Comment (RomanDavlyatshin) [reply common/src/main/kotlin/AgentLoader.kt:51]:
```
    fun downloadAgent(
        config: Configuration,
        dir: File
    ) = run {
        var zipPath: File
        if (config.directLocalPathToZip != null) {
            zipPath = dir.resolve(config.directLocalPathToZip)
        } else {
            val url = config.directUrlToZip ?: getDefaultRepositoryUrl(config)
            val bytes = URL(url).openStream().readBytes()
            zipPath = dir.resolve(getArtifactName(config))
            zipPath.createNewFile()
            zipPath.writeBytes(bytes)
        }
        unzip(zipPath, dir)
    }

    private fun  getDefaultRepositoryUrl (config: Configuration) =
        "${GITHUB_URL.replace("api.", "")}/$DRILL_OWNER/" +
                "${config.repositoryName}/releases/download/v" +
                "${config.version}/${getArtifactName(config)}"
    private fun getArtifactName (config: Configuration) = "${config.artifactId}-$presetName-${config.version}.zip" }
```

**Review by RomanDavlyatshin — CHANGES_REQUESTED** [review]

File: common/src/main/kotlin/AgentLoader.kt [hunk 3]
```diff
@@ -23,15 +23,40 @@ object AgentLoader {
         config: Configuration,
         dir: File
     ) = run {
+        lateinit var file: File
```
[the remark this answers is not in the document]
Comment (bodyangug) [reply common/src/main/kotlin/AgentLoader.kt]: removed

File: common/src/main/kotlin/AgentLoader.kt [hunk 4]
```diff
@@ -23,15 +23,40 @@ object AgentLoader {
...
+
+        // Case 2. directUrlToZip prop is setting-up
+        val directUrlToZip = config.directUrlToZip
         val artName = "${config.artifactId}-$presetName-${config.version}.zip"
-        val artPath = "${config.repositoryName}/releases/download/v${config.version}/$artName"
-        val agent = "${GITHUB_URL.replace("api.", "")}/$DRILL_OWNER/$artPath"
-        println("Download url is: $agent")
-        val file = dir.resolve(artName).apply { createNewFile() }
-        file.writeBytes(URL(agent).openStream().readBytes())
+        if (directUrlToZip != null && directLocalPathToZip == null) {
```
[the remark this answers is not in the document]
Comment (bodyangug) [reply common/src/main/kotlin/AgentLoader.kt]: reworked

File: common/src/main/kotlin/AgentLoader.kt [hunk 5]
```diff
@@ -23,15 +23,40 @@ object AgentLoader {
...
+        lateinit var file: File
+        // Case 1. directLocalPathToZip prop is setting-up
+        // directLocalPathToZip has the highest priority.
+        val directLocalPathToZip = config.directLocalPathToZip
+        if (directLocalPathToZip != null) {
+            println("Local path is: $directLocalPathToZip")
+            file = dir.resolve(directLocalPathToZip)
+        }
+
+        // Case 2. directUrlToZip prop is setting-up
```
[the remark this answers is not in the document]
Comment (bodyangug) [reply common/src/main/kotlin/AgentLoader.kt]: removed

File: common/src/main/kotlin/AgentLoader.kt [hunk 6]
```diff
@@ -23,15 +23,40 @@ object AgentLoader {
         config: Configuration,
         dir: File
     ) = run {
+        lateinit var file: File
+        // Case 1. directLocalPathToZip prop is setting-up
+        // directLocalPathToZip has the highest priority.
+        val directLocalPathToZip = config.directLocalPathToZip
+        if (directLocalPathToZip != null) {
+            println("Local path is: $directLocalPathToZip")
```
[the remark this answers is not in the document]
Comment (bodyangug) [reply common/src/main/kotlin/AgentLoader.kt]: reworked

File: common/src/main/kotlin/AppAgentConfiguration.kt [hunk 7]
```diff
@@ -17,7 +17,6 @@ package com.epam.drill.agent.runner
 
 open class AppAgentConfiguration : Configuration() {
     override val repositoryName: String = "java-agent"
```
[the remark this answers is not in the document]
Comment (bodyangug) [reply common/src/main/kotlin/AppAgentConfiguration.kt:19]: these two lines was removed

File: common/src/main/kotlin/AgentLoader.kt (hunk 1 above)
Comment (bodyangug) [reply common/src/main/kotlin/AgentLoader.kt:41]: done

File: common/src/main/kotlin/AgentLoader.kt (hunk 2 above)
Comment (bodyangug) [reply common/src/main/kotlin/AgentLoader.kt]: removed

File: common/src/main/kotlin/AgentLoader.kt [hunk 8]
```diff
@@ -22,16 +22,30 @@ object AgentLoader {
...
+        } else {
+            val url = config.directUrlToZip ?: getDefaultRepositoryUrl(config)
+            val bytes = URL(url).openStream().readBytes()
+            zipPath = dir.resolve(getArtifactName(config))
+            zipPath.createNewFile()
+            zipPath.writeBytes(bytes)
+        }
+        unzip(zipPath, dir)
+    }.onFailure {
+        println("Catch exception. Cause: ${it.cause}. Message: ${it.message}")
```
[the remark this answers is not in the document]
Comment (RomanDavlyatshin) [reply common/src/main/kotlin/AgentLoader.kt:39]: according to kotlin docs printstacktrace is only available in kotlin native https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-throwable/

[the remark this answers is not in the document]
Comment (bodyangug) [reply common/src/main/kotlin/AgentLoader.kt:39]: done

**Review by RomanDavlyatshin — APPROVED** [review]

**Pull request merged**